File: src/algorithms/cpp/House_Robber_III.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <queue>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};


class Solution {
private:
    pair <int, int> dfs(TreeNode *node) {
        int v0 = 0, v1 = 0;
        if (node->left) {
            auto nxt = dfs(node->left);
            v1 += nxt.first;
            v0 += max(nxt.first, nxt.second);
        }
        if (node->right) {
            auto nxt = dfs(node->right);
            v1 += nxt.first;
            v0 += max(nxt.first, nxt.second);
        }
        return make_pair(v0, v1 + node->val);
    }

public:
    int rob(TreeNode *root) {
        if (root == nullptr)    return 0;
        queue <TreeNode *> q;
        pair <int, int> ret = dfs(root);
        return max(ret.first, ret.second);
    }
};


int main() {
    return 0;
}

File: src/algorithms/python/House_Robber_III.py
#!/bin/env python2
# -*- coding: utf-8 -*-


class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def dfs(self, root):
        v0, v1, nxts = 0, root.val, [root.left, root.right]
        for nxt in nxts:
            if nxt:
                _v = self.dfs(nxt)
                v0, v1 = v0 + max(_v[0], _v[1]), v1 + _v[0]
        return (v0, v1)

    def rob(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        return max(self.dfs(root)) if root else 0
